lib/IRGen/GenClass.cpp

```diff
@@ -139,42 +143,70 @@ namespace {
     // StructLayout.
     Optional<ArrayRef<SILType>> TailTypes;
 
-    // For now we always lay out resiliently-typed fields as if they
-    // were fragile.
+    // Normally, Swift only emits static metadata for a class if it has no
+    // generic ancestry and no fields with resilient value types, which
+    // require dynamic layout.
+    //
+    // However, for interop with Objective-C, where the runtime does not
+    // know how to invoke arbitrary code to initialize class metadata, we
+    // ignore resilience and emit a static layout and metadata for classes
+    // that would otherwise have static metadata, were it not for any
+    // resilient fields.
+    //
+    // This enables two things:
+    //
+    // - Objective-C can reference the class symbol by calling a static
+    //   method on it, for example +alloc, which requires the InstanceSize
+    //   to be known, except for possibly sliding ivars.
+    //
+    // - Objective-C message sends can call methods defined in categories
+    //   emitted by Swift, which again require the class metadata symbol
+    //   to have a static address.
+    //
+    // Note that we don't do this if the class is generic, has generic
+    // ancestry, or has a superclass that is itself resilient.
     bool CompletelyFragileLayout;
 
-    // Does the class metadata require dynamic initialization above and
-    // beyond what the runtime can automatically achieve?
-    //
-    // This is true if the class or any of its ancestors:
-    //   - is generic,
-    //   - is resilient,
-    //   - has a parent type which isn't emittable as a constant,
-    //   - or has a field with resilient layout.
-    bool ClassMetadataRequiresDynamicInitialization = false;
-
-    // Does the class have a fixed size up until the current point?
-    // If not, we have to access stored properties either ivar offset globals,
-    // or through the field offset vector, based on whether the layout has
-    // dependent layout.
-    bool ClassIsFixedSize = true;
-
-    // Does the class layout depend on generic parameters?
+    // Does the class or any of its superclasses have stored properties that
+    // where dropped due to the Swift language version availability of
+    // their types?
+    bool ClassHasMissingMembers = false;
+
+    // Does the class or any of its superclasses have stored properties of
+    // concrete type whose size is not known at compile time? This is a
+    // weaker condition than the class itself being resilient or having
+    // resilient ancestry, because we still have a fixed layout for the
+    // class metadata in this case.
+    bool ClassHasResilientMembers = false;
```

Comment:
This comment doesn't make it totally clear whether this flag is set when the class actually is resilient or has resilient ancestry.

Reply:
Updated the comment.